### PR TITLE
Fix direct memory leaks in deterministic-unit_test_drivers

### PR DESCRIPTION
**Pull request by quantumsteve**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes

Describe what this PR changes and why.  If it closes an issue, link to it here
with [a supported keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

This fixes  direct leaks found in deterministic-unit_test_hamiltonian_force by clang address sanitizer.

This is part of #3083.

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- n/a. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- n/a. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/WaveFunctionTester.cpp [hunk 1]
```diff
@@ -158,7 +158,7 @@ void WaveFunctionTester::runCloneTest()
     int nat = W.getTotalNum();
     MCWalkerConfiguration::PropertyContainer_t Properties(0,0,1,WP::MAXPROPERTIES);
     //pick the first walker
-    MCWalkerConfiguration::Walker_t* awalker = *(W.begin());
+    MCWalkerConfiguration::Walker_t* awalker = W.begin()->get();
```
Comment (ye-luo) [inline src/QMCDrivers/WaveFunctionTester.cpp:161]: Can we change pointer to const reference or at least reference?

**Review by ye-luo — COMMENTED** [review]
Could you clang-format the touched files?
Could you paste the test output of 4-4 short test cases for both CPU and legacy CUDA build?
Since MPI was touched, be a bit more cautious. We need to compare with the nightly runs.

**Comment by quantumsteve** [comment]
> Could you paste the test output of 4-4 short test cases for both CPU and legacy CUDA build?

Do you have a ctest regex in mind?

**Comment by ye-luo** [comment]
`ctest -j32 -R 4-4 -E long` try this

**Comment by quantumsteve** [comment]
```
The following tests FAILED:
	247 - short-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
	358 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
	359 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
	363 - short-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
	847 - short-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
	851 - short-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
	855 - short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check (Failed)
	859 - short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
```

**Comment by quantumsteve** [comment]
I switched back to the `develop` branch and the same tests are failing 😕  Does a specific package need to be found or option enabled?

```
The following tests FAILED:
	247 - short-LiH_ae-vmc_hf_noj_estimator_energydensity_voronoi-4-4-check (Failed)
	358 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4 (Failed)
	359 - short-LiH_pp-vmc_hf_sdj_estimator_spindensity-4-4-check (Failed)
	363 - short-LiH_pp-vmc_hf_sdj_estimator_energydensity_voronoi-4-4-check (Failed)
	847 - short-diamondC_1x1x1_pp-vmc-noJ-estimator-1rdm-4-4-check (Failed)
	851 - short-diamondC_1x1x1_pp-vmc-J2-estimator-1rdm-4-4-check (Failed)
	855 - short-diamondC_1x1x1_pp-vmc-estimator-energydensity-cell-4-4-check (Failed)
	859 - short-diamondC_1x1x1_pp-dmc-estimator-energydensity-cell-4-4-check (Failed)
```

**Comment by ye-luo** [comment]
Since it is no worse, I think this PR is OK. The results are consistent with the nightly.
So please run clang-format and then I will approve.

**Comment by quantumsteve** [comment]
@ye-luo I ran `clang-format` version 10.0 on all of my changes. Is that the wrong version?

File: src/QMCWaveFunctions/Fermion/DiracDeterminantCUDA.cpp [hunk 2]
```diff
@@ -195,7 +195,7 @@ void DiracDeterminantCUDA::update(MCWalkerConfiguration* W,
     kdelay = W->getkDelay();
   if (kdelay > 1)
   {
-    std::vector<Walker_t*>& allwalkers = W->WalkerList;
+    auto& allwalkers                   = W->WalkerList;
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/DiracDeterminantCUDA.cpp:198]: strange space

File: src/QMCDrivers/WFOpt/HamiltonianRef.cpp [hunk 3]
```diff
@@ -75,7 +75,7 @@ int HamiltonianRef::addObservables(ParticleSet& P)
 void HamiltonianRef::evaluate(MCWalkerConfiguration& W, std::vector<RealType>& energyVector)
 {
   using WP = WalkerProperties::Indexes;
-  std::vector<Walker_t*>& walkers = W.WalkerList;
+  auto& walkers                   = W.WalkerList;
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/HamiltonianRef.cpp:78]: strange space.

**Review by ye-luo — COMMENTED** [review]
I saw a few places not seem well formatted.

**Comment by prckent** [comment]
@quantumsteve I think these are tests that always fail because they are not fully implemented. The errors are "correct". If you ran with output on failure, check the output error message.  Other than the deterministic tests, we have not gone through (yet) and fully cleaned up all working/not-working combinations. It is very sad that the spin-density is not fully implemented+working but that is the situation. 

When it comes to running tests on (say) sulfur, `ctest -j 16 -E long` only takes a couple of hours. This is my standard protocol for any non-trivial change. If you configure QMC_DATA you might want to also exclude some or all of the performance tests since these have no meaningful science-based error check other than if they crash.

File: src/QMCDrivers/WFOpt/HamiltonianRef.cpp (hunk 3 above)
Comment (prckent) [reply src/QMCDrivers/WFOpt/HamiltonianRef.cpp:78]: It is part of a two line list with the = aligned in the same column.

File: src/QMCHamiltonians/QMCHamiltonian.cpp [hunk 4]
```diff
@@ -1019,7 +1019,7 @@ QMCHamiltonian* QMCHamiltonian::makeClone(ParticleSet& qp, TrialWaveFunction& ps
 void QMCHamiltonian::evaluate(MCWalkerConfiguration& W, std::vector<RealType>& energyVector)
 {
   ScopedTimer local_timer(ham_timer_);
-  std::vector<Walker_t*>& walkers = W.WalkerList;
+  auto& walkers                   = W.WalkerList;
```
[the remark this answers is not in the document]
Comment (quantumsteve) [reply src/QMCHamiltonians/QMCHamiltonian.cpp:1022]: Looks like `git clang-format` doesn't change nearby lines. I push a fix in a minute.

[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCHamiltonians/QMCHamiltonian.cpp:1022]: clang 11 works fine.

**Comment by ye-luo** [comment]
Test this please

**Review by ye-luo — APPROVED** [review]

**Pull request merged**